### PR TITLE
[Pickles] Pass previous proof inputs from circuit logic

### PR DESCRIPTION
**Pull request by mrmr1993**
This PR builds upon #11016, tweaking the behaviour of the circuit's `main` function to return the public inputs of the previous proofs instead of consuming them.

Since this changes the external API of pickles, all circuits in the codebase had to be updated.

Checklist:

- [ ] Modified the current draft of release notes with details on what is completed or incomplete within this project
- [ ] Document code purpose, how to use it
  - Mention expected invariants, implicit constraints
- [ ] Tests were added for the new behavior
  - Document test purpose, significance of failures
  - Test names should reflect their purpose
- [ ] All tests pass (CI will check this if you didn't)
- [ ] Serialized types are in stable-versioned modules
- [ ] Does this close issues? List them

### CODE REVIEW

**Review by mitschabaude — APPROVED** [review]
snarkyjs change gets my 👍🏻

File: src/lib/blockchain_snark/blockchain_snark_state.ml [hunk 1]
```diff
@@ -86,26 +91,40 @@ let non_pc_registers_equal_var t1 t2 =
 let%snarkydef step ~(logger : Logger.t)
     ~(proof_level : Genesis_constants.Proof_level.t)
     ~(constraint_constants : Genesis_constants.Constraint_constants.t)
-    Hlist.HlistId.
-      [ previous_state_hash
-      ; (txn_snark : Transaction_snark.Statement.With_sok.Checked.t)
-      ] new_state_hash : _ Tick.Checked.t =
+    new_state_hash : _ Tick.Checked.t =
```
Comment (deepthiskumar) [inline src/lib/blockchain_snark/blockchain_snark_state.ml:94]: just `new_state`?

File: src/lib/blockchain_snark/blockchain_snark_state.ml [hunk 2]
```diff
@@ -280,61 +299,54 @@ let%snarkydef step ~(logger : Logger.t)
...
+  let to_field_elements (t : t) = [| Data_as_hash.hash t |]
+end
+
+let typ = Data_as_hash.typ ~hash:(fun t -> (Protocol_state.hashes t).state_hash)
+
+type tag = (Statement_var.t, Statement.t, Nat.N2.n, Nat.N1.n) Pickles.Tag.t
 
-let check w ?handler ~proof_level ~constraint_constants txn_snark new_state_hash
-    : unit Or_error.t =
+let check w ?handler ~proof_level ~constraint_constants new_state_hash :
```
Comment (deepthiskumar) [inline src/lib/blockchain_snark/blockchain_snark_state.ml:327]: `new_state`?

**Review by deepthiskumar — APPROVED** [review]
Blockchain snark, transaction snark, zkapp examples :heavy_check_mark:

File: src/lib/blockchain_snark/blockchain_snark_state.ml [hunk 3]
```diff
@@ -86,26 +91,40 @@ let non_pc_registers_equal_var t1 t2 =
...
+      (exists Transaction_snark.Statement.With_sok.typ
+         ~request:(As_prover.return Txn_snark) )
+  in
+  let%bind ( previous_state
+           , previous_state_hash
+           , previous_blockchain_proof_input
+           , previous_state_body_hash ) =
+    let%bind prev_state_ref =
+      with_label __LOC__
+        (exists (Typ.Internal.ref ()) ~request:(As_prover.return Prev_state))
```
Comment (deepthiskumar) [inline src/lib/blockchain_snark/blockchain_snark_state.ml:113]: why are we using `Typ.Internal.ref ()` and not `Protocol_state.typ`?

Comment (mrmr1993) [reply src/lib/blockchain_snark/blockchain_snark_state.ml:113]: We need the `As_prover.Ref.t` for `previous_blockchain_proof_input` anyway, it ends up being less code to do it this way.

**Pull request merged**